horovod/common/controller.cc

```diff
@@ -324,7 +333,7 @@ ResponseList Controller::ComputeResponseList(std::atomic_bool& shut_down,
       if ((response.response_type() == Response::ResponseType::ALLREDUCE ||
            response.response_type() == Response::ResponseType::ADASUM) &&
           (int)response.devices().size() == size_) {
-        response_cache_.put(response, tensor_queue_);
+        response_cache_.put(response, tensor_queue_, state.joined);
```

Comment:
I think we need to be careful here in the cache handling. I see that if you have this `state.joined` flag set, the cache doesn't save any of the parameters from the tensor table entry that are not normally stored within the responses (`device`, `dtype`, `shape`), since you know that the tensors won't be in the table. This is tricky because it now puts the cache in an inconsistent state across workers (for a particular response, some workers will have the additional tensor information, while others will not). We need to make sure that when the join operation is concluded, the cache is in a globally consistent state across all workers. 

For example, will this work in the following scenarios? (assume there are two ranks, either joining or submitting tensor A, each tuple is one set of operations submitted):
1. (join, A), (join, join) (A, A)
2. (A, A), (join, A), (join, join), (A, A)

The difference between the two cases is that in the first case, worker 0 does not have a preexisting cache entry for A, while in case 2, worker 0 does have a prexisting cached entry. 

On a related note, I can't seem to find where `state.joined` is reset back to false?

Reply:
`state.joined` reset back to false on execute:
https://github.com/horovod/horovod/blob/master/horovod/common/ops/collective_operations.cc#L222

Reply:
> (join, A), (join, join) (A, A)
> (A, A), (join, A), (join, join), (A, A)

I think I need to explain the Join op more.

After a rank sends Join request, it does not send any horovod requests (can still communicate with the coordinator, but with empty request list) until it gets Join response (which is sent by the coordinator after all ranks sent Join request). After getting Join response, every rank reset their Join state and continue as normal.

So "(join, A), (join, join) (A, A)" is not possible, but "(join, A), (-, join), (A, A)" is possible, there will be communication of every of these 3 steps.

"(A, A), (join, A), (join, join), (A, A)" will be "(A, A), (join, A), (-, join), (A, A)". Second step will use cache, and on the 0-th rank the cache info for the tensor A will be updated: `params.joined = true` and params.shape will be empty https://github.com/horovod/horovod/pull/1571/files#diff-9b02a25ea514cc98deff8de867ffd54dR177. So on the forth step the shape of the A tensor in the 0-th rank cache will be different from the shape in the request, so cache invalidation will happen.

Reply:
Your notation is much clearer. I was just trying to indicate that a rank was still in a join state, not that it'd repeatedly send join requests.

So to focus on the `(A, A), (join, A), (-, join), (A, A)` case:
- "Second step will use cache, and on the 0-th rank the cache info for the tensor A will be updated: `params.joined = true` and params.shape will be empty" <-- Does this operate correctly given this block of code? https://github.com/horovod/horovod/blob/9723db4302d8692a0c2631a0c4c83d4be70e002b/horovod/common/response_cache.cc#L102
In the `put_` method, the existing cache entry is just moved if the cache state is HIT. Since A already exists in the cache, I don't think the params will be updated. 
- I don't see how the `params.joined = true` state is reset. On the cycle that rank 1 submits the join, at the time the responses are recached here:
https://github.com/horovod/horovod/blob/9723db4302d8692a0c2631a0c4c83d4be70e002b/horovod/common/controller.cc#L336
I think both workers will have `state.joined = true` still right, since the state isn't reset until the after the operation is executed?